### PR TITLE
MAILET-115 Ensure AbstractNotify/AbstractRedirect has a good coverage and meet new standards

### PR DESCRIPTION
**Pull request by aduprat**

### CODE REVIEW

File: mailet/base/src/test/java/org/apache/mailet/base/test/FakeMail.java [hunk 1]
```diff
@@ -198,57 +264,58 @@ public void setMessage(MimeMessage message) {
...
         this.size = size;
     }
-
-    public void setRemoteAddr(String remoteAddr) {
-        this.remoteAddr = remoteAddr;
-    }
-
-    public void setSender(MailAddress sender) {
-        this.sender = sender;
-    }
```
Comment (rouazana) [inline mailet/base/src/test/java/org/apache/mailet/base/test/FakeMail.java]: Why did you remove these methods? They can be useful in some tests (and I use setRemoteAddr in my PR ^^)

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/OnlyTextTest.java [hunk 2]
```diff
@@ -88,7 +90,9 @@ public void testService() throws MessagingException, IOException {
         message.setContent(mp);
         message.saveChanges();
 
-        mail = new FakeMail(message);
+        mail = FakeMail.builder()
+                .mimeMessage(message)
+                .build();
```
Comment (rouazana) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/OnlyTextTest.java]: In these (test) cases, I'm not sure the builder is more clear neither more simple to write

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/send/MailSpoolTest.java [hunk 3]
```diff
@@ -54,7 +54,7 @@ public MailQueue getQueue(String name) {
 
     @Test
     public void sendShouldEnQueueTheMail() throws Exception {
-        FakeMail mail = new FakeMail();
+        FakeMail mail = FakeMail.builder().build();
```
Comment (rouazana) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/send/MailSpoolTest.java]: Idem, what do you think of keeping this constructor when it helps?

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/BounceTest.java [hunk 4]
```diff
@@ -58,30 +65,104 @@ public void setUp() throws Exception {
...
+        mimeMessage.setText("My content");
+        mail.setMessage(mimeMessage);
+
+        bounce.service(mail);
+
+        assertThat(mail.getState()).isEqualTo(Mail.GHOST);
+    }
+
+    @Test
+    public void bounceShouldNotChandeTheStateWhenNoSenderAndTrue() throws Exception {
```
Comment (rouazana) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/BounceTest.java]: s/Chande/Change/

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/BounceTest.java [hunk 5]
```diff
@@ -58,30 +65,104 @@ public void setUp() throws Exception {
...
 
         bounce.service(mail);
 
-        FakeMailContext.SentMail expected = new FakeMailContext.SentMail(null,
-            Lists.newArrayList(senderMailAddress),
-            null,
-            ImmutableMap.<String, Serializable>of());
+        MailAddress sender = null;
+        Collection<MailAddress> recipients = Lists.newArrayList(senderMailAddress);
+        MimeMessage message = null;
```
Comment (mbaechler) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/BounceTest.java]: why a null message ?

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/BounceTest.java [hunk 6]
```diff
@@ -58,30 +65,104 @@ public void setUp() throws Exception {
...
+        Map<String, Serializable> attributes = ImmutableMap.<String, Serializable> of();
+        FakeMailContext.SentMail expected = new FakeMailContext.SentMail(sender, recipients, message, attributes);
         assertThat(fakeMailContext.getSentMails()).containsOnly(expected);
     }
 
+    @Test
+    public void bounceShouldChangeTheStateWhenNoSenderAndFalse() throws Exception {
```
Comment (mbaechler) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/BounceTest.java]: andFalse ?

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/DSNBounceTest.java [hunk 7]
```diff
@@ -0,0 +1,404 @@
...
+        dsnBounce = new DSNBounce();
+        DNSService dnsService = mock(DNSService.class);
+        dsnBounce.setDNSService(dnsService);
+        fakeMailContext = FakeMailContext.defaultContext();
+
+        InetAddress localHost = InetAddress.getLocalHost();
+        when(dnsService.getLocalHost())
+            .thenReturn(localHost);
+        when(dnsService.getHostName(localHost))
+            .thenReturn("My host");
```
Comment (mbaechler) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/DSNBounceTest.java]: I think hostname can't contain spaces

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/DSNBounceTest.java [hunk 8]
```diff
@@ -0,0 +1,404 @@
...
+
+        dsnBounce.service(mail);
+
+        List<SentMail> sentMails = fakeMailContext.getSentMails();
+        assertThat(sentMails).hasSize(1);
+        SentMail sentMail = sentMails.get(0);
+        MimeMessage sentMessage = sentMail.getMsg();
+        MimeMultipartReport content = (MimeMultipartReport) sentMessage.getContent();
+        String hostname = InetAddress.getLocalHost().getHostName();
+        assertThat(content.getBodyPart(0).getContent()).isEqualTo("Hi. This is the James mail server at " + hostname + ".\nI'm afraid I wasn't able to deliver your message to the following addresses.\nThis is a permanent error; I've given up. Sorry it didn't work out.  Below\nI include the list of recipients and the reason why I was unable to deliver\nyour message.\n\n" +
```
Comment (mbaechler) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/DSNBounceTest.java]: the assertion is too accurate and doesn't match test title

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/DSNBounceTest.java [hunk 9]
```diff
@@ -0,0 +1,404 @@
...
+
+    @Test
+    public void serviceShouldSendMultipartMailWithDSNPart() throws Exception {
+        FakeMailetConfig mailetConfig = new FakeMailetConfig(MAILET_NAME, fakeMailContext);
+        dsnBounce.init(mailetConfig);
+
+        MailAddress senderMailAddress = new MailAddress("sender@domain.com");
+        MimeMessage mimeMessage = new MimeMessage(Session.getDefaultInstance(new Properties()));
+        mimeMessage.setText("My content");
+        Date lastUpdated = new Date();
```
Comment (mbaechler) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/DSNBounceTest.java]: please use a fixed date

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/DSNBounceTest.java [hunk 10]
```diff
@@ -0,0 +1,404 @@
...
+        List<SentMail> sentMails = fakeMailContext.getSentMails();
+        assertThat(sentMails).hasSize(1);
+        SentMail sentMail = sentMails.get(0);
+        assertThat(sentMail.getSender()).isNull();
+        assertThat(sentMail.getRecipients()).containsOnly(senderMailAddress);
+        MimeMessage sentMessage = sentMail.getMsg();
+        MimeMultipartReport content = (MimeMultipartReport) sentMessage.getContent();
+
+        // Should fail to retrieve the original mail part
+        expectedException.expect(ArrayIndexOutOfBoundsException.class);
```
Comment (mbaechler) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/DSNBounceTest.java]: you can't test the number of part instead ?

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/DSNBounce.java [hunk 11]
```diff
@@ -88,112 +86,88 @@
...
-            newMail.setMessage(newMessage);
+            newMail.setMessage(createBounceMessage(originalMail));
 
             // Set additional headers
             setRecipients(newMail, getRecipients(originalMail), originalMail);
             setTo(newMail, getTo(originalMail), originalMail);
             setSubjectPrefix(newMail, getSubjectPrefix(originalMail), originalMail);
             if (newMail.getMessage().getHeader(RFC2822Headers.DATE) == null) {
-                newMail.getMessage().setHeader(RFC2822Headers.DATE, rfc822DateFormat.format(new Date()));
+                newMail.getMessage().setHeader(RFC2822Headers.DATE, RFC8822_DATE_FORMAT.format(new Date()));
```
Comment (mbaechler) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/DSNBounce.java]: how do you test that date ?

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/ResendTest.java [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+import org.apache.james.dnsservice.api.DNSService;
+import org.apache.mailet.MailAddress;
+import org.apache.mailet.base.test.FakeMailContext;
+import org.apache.mailet.base.test.FakeMailetConfig;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+
+public class ResendTest {
```
Comment (mbaechler) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/ResendTest.java]: is don't look like it's really covered by tests

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/AbstractRedirect.java [hunk 13]
```diff
@@ -152,6 +153,8 @@
 
 public abstract class AbstractRedirect extends GenericMailet {
 
+    protected abstract boolean isNotifyMailet();
```
Comment (mbaechler) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/AbstractRedirect.java]: is it temporary ?

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/AbstractRedirect.java [hunk 14]
```diff
@@ -404,9 +419,74 @@ protected String getMessage() {
      * @return {@link #getMessage()}
      */
     protected String getMessage(Mail originalMail) throws MessagingException {
+        if (isNotifyMailet()) {
+            return getMessageForNotifyMailets(originalMail);
+        }
         return (isStatic()) ? this.messageText : getMessage();
     }
 
+    private String getMessageForNotifyMailets(Mail originalMail) throws MessagingException {
```
Comment (mbaechler) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/AbstractRedirect.java]: can't you extract to a delegate class right now ?

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/AbstractRedirect.java [hunk 15]
```diff
@@ -812,6 +910,18 @@ protected String getSubjectPrefix(Mail originalMail) throws MessagingException {
      * method.
      */
     protected void setSubjectPrefix(Mail newMail, String subjectPrefix, Mail originalMail) throws MessagingException {
+        if (isNotifyMailet()) {
+            String subject = originalMail.getMessage().getSubject();
```
Comment (mbaechler) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/AbstractRedirect.java]: there is a mailet for that, we should reuse the code there

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/AbstractRedirect.java [hunk 16]
```diff
@@ -153,41 +155,61 @@
...
-     * Controls certain log messages.
-     */
     protected boolean isDebug = false;
 
-    /**
-     * Holds the value of the <code>static</code> init parameter.
-     */
     protected boolean isStatic = false;
 
+    private static enum SpecialAddressKind {
```
Comment (mbaechler) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/AbstractRedirect.java]: what about having a toMailAddress method here ?

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/AbstractRedirect.java [hunk 17]
```diff
@@ -427,64 +437,67 @@ protected String getMessage(Mail originalMail) throws MessagingException {
 
     private String getMessageForNotifyMailets(Mail originalMail) throws MessagingException {
         MimeMessage message = originalMail.getMessage();
-        StringWriter sout = new StringWriter();
-        PrintWriter out = new PrintWriter(sout, true);
+        StringBuffer buffer = new StringBuffer();
```
Comment (mbaechler) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/AbstractRedirect.java]: I guess you rather want a StringBuilder

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/AbstractRedirect.java [hunk 18]
```diff
@@ -427,64 +437,67 @@ protected String getMessage(Mail originalMail) throws MessagingException {
...
 
-        // First add the "local" notice
-        // (either from conf or generic error message)
-        out.println(getMessage());
-        // And then the message from other mailets
+        buffer.append(getMessage()).append(LINE_BREAK);
```
Comment (mbaechler) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/AbstractRedirect.java]: what is the point using a StringBuffer if you have to code println by hand ?

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/AbstractRedirect.java [hunk 19]
```diff
@@ -617,22 +622,19 @@ protected void setRecipients(Mail newMail, Collection<MailAddress> recipients, M
     protected InternetAddress[] getTo(Mail originalMail) throws MessagingException {
```
Comment (mbaechler) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/AbstractRedirect.java]: why do we need to keep this raw array on a protected method ?

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/AbstractRedirect.java [hunk 20]
```diff
@@ -617,22 +622,19 @@ protected void setRecipients(Mail newMail, Collection<MailAddress> recipients, M
...
             }
+            Collection<InternetAddress> addresses = replaceInternetAddresses(originalMail, ImmutableList.copyOf(apparentlyTo));
+            return addresses.toArray(new InternetAddress[addresses.size()]);
         }
 
-        return apparentlyTo;
+        return null;
+    }
+
+    private boolean containsOnlyUnalteredOrTo(InternetAddress[] to) {
```
Comment (mbaechler) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/AbstractRedirect.java]: what about using something else than raw array ?

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/AbstractRedirect.java [hunk 21]
```diff
@@ -1510,24 +1511,29 @@ public static void changeSubject(MimeMessage message, String newValue) throws Me
      *                header was not present (in this case it always return null).
      * @return the MIME charset name or null if no encoding applied
      */
-    static private String determineMailHeaderEncodingCharset(String rawText) {
-        if (rawText == null)
+    @VisibleForTesting String determineMailHeaderEncodingCharset(String rawText) {
```
Comment (mbaechler) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/AbstractRedirect.java]: should be in a shared class, we already have the same piece of code elsewhere (another mailet I rewrote)

**Comment by mbaechler** [comment]
there's still work to do on AbstractRedirect, don't you think ? maybe some pair-programming would help here

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/DSNBounceTest.java [hunk 22]
```diff
@@ -0,0 +1,404 @@
...
+                .remoteAddr("remoteHost")
+                .build();
+
+        dsnBounce.service(mail);
+
+        assertThat(mail.getState()).isNull();
+    }
+
+    @Test
+    public void serviceShouldNotAttachedTheOriginalMailWhenAttachmentIsEqualToNone() throws Exception {
```
Comment (rouazana) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/DSNBounceTest.java]: s/Attached/Attach/

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/DSNBounceTest.java [hunk 23]
```diff
@@ -0,0 +1,404 @@
...
+        MimeMessage sentMessage = sentMail.getMsg();
+        MimeMultipartReport content = (MimeMultipartReport) sentMessage.getContent();
+
+        // Should fail to retrieve the original mail part
+        expectedException.expect(ArrayIndexOutOfBoundsException.class);
+        content.getBodyPart(2);
+    }
+
+    @Test
+    public void serviceShouldAttachedTheOriginalMailWhenAttachmentIsEqualToAll() throws Exception {
```
Comment (rouazana) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/DSNBounceTest.java]: idem

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/DSNBounceTest.java [hunk 24]
```diff
@@ -0,0 +1,404 @@
...
+        SentMail sentMail = sentMails.get(0);
+        assertThat(sentMail.getSender()).isNull();
+        assertThat(sentMail.getRecipients()).containsOnly(senderMailAddress);
+        MimeMessage sentMessage = sentMail.getMsg();
+        MimeMultipartReport content = (MimeMultipartReport) sentMessage.getContent();
+        assertThat(content.getBodyPart(2).getContent()).isEqualTo(mimeMessage);
+    }
+
+    @Test
+    public void serviceShouldAttachedTheOriginalMailHeadersOnlyWhenAttachmentIsEqualToHeads() throws Exception {
```
Comment (rouazana) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/DSNBounceTest.java]: idem

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/NotifySenderTest.java [hunk 25]
```diff
@@ -0,0 +1,160 @@
...
+    public void getToShouldReturnFromWhenToIsEqualToFrom() throws Exception {
+        FakeMailetConfig mailetConfig = new FakeMailetConfig(MAILET_NAME, fakeMailContext);
+        mailetConfig.setProperty("to", "from");
+        notifySender.init(mailetConfig);
+
+        assertThat(notifySender.getTo()).containsOnly(SpecialAddress.FROM.toInternetAddress());
+    }
+
+    @Test
+    public void getToShouldReturnSenderWhenToIsNotEqualToSenderUnalteredOfFrom() throws Exception {
```
Comment (rouazana) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/NotifySenderTest.java]: s/Of/OR/ ?

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/ForwardTest.java [hunk 26]
```diff
@@ -0,0 +1,199 @@
...
+    }
+
+    @Test
+    public void initShouldNotThrowWhenEveryParameters() throws Exception {
+        FakeMailetConfig mailetConfig = new FakeMailetConfig(MAILET_NAME, fakeMailContext);
+        mailetConfig.setProperty("debug", "true");
+        mailetConfig.setProperty("passThrough", "false");
+        mailetConfig.setProperty("fakeDomainCheck", "false");
+        mailetConfig.setProperty("forwardto", "other@james.org");
+        mailetConfig.setProperty("forwardTo", "other@james.org");
```
Comment (rouazana) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/ForwardTest.java]: is it expected to have the same parameter twice?

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/RedirectTest.java [hunk 27]
```diff
@@ -0,0 +1,351 @@
...
+        mailetConfig.setProperty("debug", "true");
+        mailetConfig.setProperty("passThrough", "false");
+        mailetConfig.setProperty("fakeDomainCheck", "false");
+        mailetConfig.setProperty("inline", "true");
+        mailetConfig.setProperty("attachment", "true");
+        mailetConfig.setProperty("message", "mess");
+        mailetConfig.setProperty("recipients", "user@james.org, user2@james.org");
+        mailetConfig.setProperty("to", "to@james.org");
+        mailetConfig.setProperty("replyTo", "replyTo@james.org");
+        mailetConfig.setProperty("replyto", "replyto@james.org");
```
Comment (rouazana) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/RedirectTest.java]: twice?

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/RedirectTest.java [hunk 28]
```diff
@@ -0,0 +1,351 @@
...
+        mailetConfig.setProperty("subject", "subj");
+        mailetConfig.setProperty("prefix", "pref");
+        mailetConfig.setProperty("attachError", "true");
+        mailetConfig.setProperty("isReply", "true");
+
+        redirect.init(mailetConfig);
+    }
+
+    @Test
+    public void initShouldReturnNullWhenNoRecipientsOrToToParameters() throws Exception {
```
Comment (rouazana) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/RedirectTest.java]: ToTo?

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/Redirect.java [hunk 29]
```diff
@@ -281,142 +285,121 @@
...
+        }
 
-            try {
-                return new MailAddress(addressString);
-            } catch (Exception e) {
-                throw new MessagingException("Exception thrown in getReversePath() parsing: " + addressString, e);
-            }
+        MailAddress specialAddress = getSpecialAddress(addressString, new String[] { "postmaster", "sender", "null" });
+        if (specialAddress != null) {
+            return specialAddress;
```
Comment (rouazana) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/Redirect.java]: Can't you modify getSpecialAddress to take a default value as last parameter? You use this construct twice.

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/ResendTest.java [hunk 30]
```diff
@@ -0,0 +1,100 @@
...
+        when(dnsService.getLocalHost()).thenThrow(new UnknownHostException());
+    }
+
+    @Test
+    public void getMailetInfoShouldReturnValue() {
+        assertThat(resend.getMailetInfo()).isEqualTo("Redirect Mailet");
+    }
+
+    @Test
+    public void initShouldThrowWhenUnkownParameter() throws Exception {
```
Comment (rouazana) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/ResendTest.java]: s/k/kn

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/AbstractRedirectTest.java [hunk 31]
```diff
@@ -0,0 +1,517 @@
...
+
+        MailAddress address = new MailAddress("user", "address.marker");
+        MailAddress address2 = new MailAddress("user2", "address.marker");
+        Collection<MailAddress> addresses = testee.replaceMailAddresses(mail, ImmutableList.of(address, address2));
+
+        assertThat(addresses).containsOnly(address, address2);
+    }
+
+    @Test
+    public void replaceMailAddressesShouldSameListWhenAddressesMatchesDelete() throws Exception {
```
Comment (rouazana) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/AbstractRedirectTest.java]: Should what?

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/AbstractRedirectTest.java [hunk 32]
```diff
@@ -0,0 +1,517 @@
...
+        InternetAddress internetAddress2 = new InternetAddress("user2@address.mar");
+        ImmutableList<InternetAddress> list = ImmutableList.of(internetAddress, internetAddress2);
+
+        Collection<InternetAddress> addresses = testee.replaceInternetAddresses(mail, list);
+
+        assertThat(addresses).containsOnly(internetAddress, internetAddress2);
+    }
+
+    @Test
+    public void replaceInternetAddressesShouldReturnSenderWhenAddressesMatchesSender() throws Exception {
```
Comment (rouazana) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/AbstractRedirectTest.java]: s/Matched/Match/ ?

Comment (rouazana) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/AbstractRedirectTest.java]: and below (or s/Addresses/Address/)

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/BounceTest.java [hunk 33]
```diff
@@ -58,30 +65,104 @@ public void setUp() throws Exception {
...
+        FakeMailetConfig mailetConfig = new FakeMailetConfig(MAILET_NAME, fakeMailContext);
```
Comment (chibenwa) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/BounceTest.java]: Does a mail is returned to sender when pass through is false ?

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/Bounce.java [hunk 34]
```diff
@@ -103,77 +106,55 @@
...
     @Override
     public void service(Mail originalMail) throws MessagingException {
         if (originalMail.getSender() == null) {
-            if (isDebug)
-                log("Processing a bounce request for a message with an empty reverse-path.  No bounce will be sent.");
-            if (!getPassThrough(originalMail)) {
-                originalMail.setState(Mail.GHOST);
+            passThrough(originalMail);
+        } else {
+            if (isDebug) {
```
Comment (chibenwa) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/Bounce.java]: Do we really need the is debug parameter ? Can't this be handled by the Logger directly ?

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/AbstractRedirect.java [hunk 35]
```diff
@@ -502,33 +515,40 @@ private String getMessageForNotifyMailets(Mail originalMail) throws MessagingExc
...
-            InternetAddress[] iaarray = InternetAddress.parse(addressList, false);
-            for (InternetAddress anIaarray : iaarray) {
-                String addressString = anIaarray.getAddress();
-                MailAddress specialAddress = getSpecialAddress(addressString, new String[]{"postmaster", "sender", "from", "replyTo", "reversePath", "unaltered", "recipients", "to", "null"});
-                if (specialAddress != null) {
-                    newRecipients.add(specialAddress);
-                } else {
-                    newRecipients.add(new MailAddress(anIaarray));
-                }
+            MailAddress specialAddress = getSpecialAddress(address.getAddress(), allowedSpecials);
```
Comment (chibenwa) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/AbstractRedirect.java]: Optional ?

File: mailet/base/src/test/java/org/apache/mailet/base/test/FakeMail.java (hunk 1 above)
Comment (aduprat) [inline mailet/base/src/test/java/org/apache/mailet/base/test/FakeMail.java]: Ok, i will reinject them; i delete them because they were not used and not overriding something.

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/OnlyTextTest.java (hunk 2 above)
Comment (aduprat) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/OnlyTextTest.java]: Ok, i added a static method for that.

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/send/MailSpoolTest.java (hunk 3 above)
Comment (aduprat) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/send/MailSpoolTest.java]: same

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/BounceTest.java (hunk 4 above)
Comment (aduprat) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/BounceTest.java]: OK

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/BounceTest.java (hunk 5 above)
Comment (aduprat) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/BounceTest.java]: Because we are using FakeMailContext which equals method does not check message equality;
and there's no equals method in java mail MimeMessage class.

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/BounceTest.java (hunk 6 above)
Comment (aduprat) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/BounceTest.java]: OK

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/DSNBounceTest.java (hunk 7 above)
Comment (aduprat) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/DSNBounceTest.java]: right

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/DSNBounceTest.java (hunk 8 above)
Comment (aduprat) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/DSNBounceTest.java]: OK

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/DSNBounceTest.java (hunk 9 above)
Comment (aduprat) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/DSNBounceTest.java]: OK

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/DSNBounceTest.java (hunk 10 above)
Comment (aduprat) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/DSNBounceTest.java]: Ok, the java mail API is not really sexy

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/DSNBounce.java (hunk 11 above)
Comment (aduprat) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/DSNBounce.java]: OK

Comment (aduprat) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/DSNBounce.java]: Some tests added, refactoring/testing SimplifiedDateFormat should be done in MAILET-132

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/ResendTest.java (hunk 12 above)
Comment (aduprat) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/ResendTest.java]: It is, i cover AbstractRedirect methods in another commit.

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/AbstractRedirect.java (hunk 13 above)
Comment (aduprat) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/AbstractRedirect.java]: Yes, still refactoring

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/AbstractRedirect.java (hunk 14 above)
Comment (aduprat) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/AbstractRedirect.java]: OK

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/AbstractRedirect.java (hunk 15 above)
Comment (aduprat) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/AbstractRedirect.java]: OK

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/AbstractRedirect.java (hunk 16 above)
Comment (aduprat) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/AbstractRedirect.java]: Not really the responsability of this enum, it's more logic to have it in AddressMarker.

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/AbstractRedirect.java (hunk 17 above)
Comment (aduprat) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/AbstractRedirect.java]: Right

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/AbstractRedirect.java (hunk 18 above)
Comment (aduprat) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/AbstractRedirect.java]: I may revert if you prefer StringWriter + PrintWriter

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/AbstractRedirect.java (hunk 19 above)
Comment (aduprat) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/AbstractRedirect.java]: I noted that in the second step of refactoring.

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/AbstractRedirect.java (hunk 20 above)
Comment (aduprat) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/AbstractRedirect.java]: I noted that in the second step of refactoring.

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/AbstractRedirect.java (hunk 21 above)
Comment (aduprat) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/AbstractRedirect.java]: Which one?

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/DSNBounceTest.java (hunk 22 above)
Comment (aduprat) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/DSNBounceTest.java]: OK

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/DSNBounceTest.java (hunk 23 above)
Comment (aduprat) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/DSNBounceTest.java]: OK

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/DSNBounceTest.java (hunk 24 above)
Comment (aduprat) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/DSNBounceTest.java]: OK

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/NotifySenderTest.java (hunk 25 above)
Comment (aduprat) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/NotifySenderTest.java]: OK

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/ForwardTest.java (hunk 26 above)
Comment (aduprat) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/ForwardTest.java]: There's no check for that, they are not the same lowercase/camelcase

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/RedirectTest.java (hunk 27 above)
Comment (aduprat) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/RedirectTest.java]: same comment

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/RedirectTest.java (hunk 28 above)
Comment (aduprat) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/RedirectTest.java]: OK

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/Redirect.java (hunk 29 above)
Comment (aduprat) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/Redirect.java]: The default is taken from init parameters, i may have a look of that in a second refactoring step.

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/ResendTest.java (hunk 30 above)
Comment (aduprat) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/ResendTest.java]: OK

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/AbstractRedirectTest.java (hunk 31 above)
Comment (aduprat) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/AbstractRedirectTest.java]: OK

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/AbstractRedirectTest.java (hunk 32 above)
Comment (aduprat) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/AbstractRedirectTest.java]: OK

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/AbstractRedirect.java (hunk 18 above)
Comment (mbaechler) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/AbstractRedirect.java]: I don't care, but printwriter is shorter because of println method

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/AbstractRedirect.java (hunk 21 above)
Comment (mbaechler) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/AbstractRedirect.java]: oh, I actually deleted the code because I solved the problem with another solution. Don't bother with this comment.

**Comment by aduprat** [comment]
Replace by #402
